### ISSUE TITLE
Possible route collision bug

### ISSUE THREAD

**Issue by dselans** [Kong/kong#2948]
### Summary

It seems that we have ran into a really bizarre route (collision?) bug.

Given two routes - `/e` and `/example` - if one does not have any methods defined on the first one, while the second has a ["GET" ...], when hitting `/example`, kong will forward the request to the upstream of `/e` with the request URI set to `/xample` (yes, without an `e`).

### Steps To Reproduce

1. Run two http servers -- one on :8080, one on :8181
2. Configure two API's in kong as follows:

```curl -i -X POST \
  --url http://localhost:8001/apis/ \
  --data 'name=example-api' \
  --data uris=/example-api \
  --data preserve_host=true \
  --data 'upstream_url=http://localhost:8080'

curl -i -X POST \
  --url http://localhost:8001/apis/ \
  --data 'name=e' \
  --data uris=/e \
  --data preserve_host=true \
  --data 'upstream_url=http://localhost:8181'\
  --data 'methods=GET'
```

3. Hit `/e` via kong; everything should work as expected. Inbound request on first server will look something like this:

```
----- Request Start ----->

/
Host: localhost
Connection: keep-alive
X-Forwarded-For: 172.18.0.1
X-Forwarded-Proto: http
X-Forwarded-Host: localhost
X-Forwarded-Port: 80
X-Real-IP: 172.18.0.1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.8,ca;q=0.6
Cookie: foo=bar

<----- Request End -----
```

4. Now hit `/example-api`; you would expect it to go to /example-api on :8080, but you get delivered to the server on :8181 (w/ `/xample-api` ???).

The output of the request on the upstream of `/e` looks something like this:

```
----- Request Start ----->

/xample-api
Host: localhost
Connection: keep-alive
X-Forwarded-For: 172.18.0.1
X-Forwarded-Proto: http
X-Forwarded-Host: localhost
X-Forwarded-Port: 80
X-Real-IP: 172.18.0.1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_6) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/61.0.3163.100 Safari/537.36
Upgrade-Insecure-Requests: 1
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.8,ca;q=0.6
Cookie: foo=bar

<----- Request End -----
```

### Additional Details & Logs

- Kong version (`$ kong version`)
    + LUAROCKS_VERSION=2.4.1
    + KONG_VERSION=0.11.0
    + SERF_VERSION=0.8.1
- Kong debug-level startup logs (`$ kong start --vv`)
```
2017/10/10 18:01:22 [debug] ngx_lua: 10008
2017/10/10 18:01:22 [debug] nginx: 1011002
2017/10/10 18:01:22 [debug] Lua: LuaJIT 2.1.0-beta2
2017/10/10 18:01:22 [verbose] reading config file at /etc/kong/kong.conf
2017/10/10 18:01:22 [debug] KONG_SSL_CERT_KEY ENV found with "/usr/local/kong/ssl/server.key"
2017/10/10 18:01:22 [debug] KONG_PG_USER ENV found with "kong"
2017/10/10 18:01:22 [debug] KONG_PG_HOST ENV found with "postgres-94"
2017/10/10 18:01:22 [debug] KONG_PROXY_LISTEN ENV found with "0.0.0.0:80"
2017/10/10 18:01:22 [debug] KONG_DATABASE ENV found with "postgres"
2017/10/10 18:01:22 [debug] KONG_PROXY_LISTEN_SSL ENV found with "0.0.0.0:443"
2017/10/10 18:01:22 [debug] KONG_SSL_CERT ENV found with "/usr/local/kong/ssl/server.crt"
2017/10/10 18:01:22 [debug] KONG_NGINX_DAEMON ENV found with "off"
2017/10/10 18:01:22 [debug] KONG_PG_PASSWORD ENV found with "******"
2017/10/10 18:01:22 [debug] admin_access_log = "logs/admin_access.log"
2017/10/10 18:01:22 [debug] admin_error_log = "logs/error.log"
2017/10/10 18:01:22 [debug] admin_http2 = false
2017/10/10 18:01:22 [debug] admin_listen = "0.0.0.0:8001"
2017/10/10 18:01:22 [debug] admin_listen_ssl = "0.0.0.0:8444"
2017/10/10 18:01:22 [debug] admin_ssl = true
2017/10/10 18:01:22 [debug] anonymous_reports = true
2017/10/10 18:01:22 [debug] cassandra_consistency = "ONE"
2017/10/10 18:01:22 [debug] cassandra_contact_points = {"127.0.0.1"}
2017/10/10 18:01:22 [debug] cassandra_data_centers = {"dc1:2","dc2:3"}
2017/10/10 18:01:22 [debug] cassandra_keyspace = "kong"
2017/10/10 18:01:22 [debug] cassandra_lb_policy = "RoundRobin"
2017/10/10 18:01:22 [debug] cassandra_port = 9042
2017/10/10 18:01:22 [debug] cassandra_repl_factor = 1
2017/10/10 18:01:22 [debug] cassandra_repl_strategy = "SimpleStrategy"
2017/10/10 18:01:22 [debug] cassandra_schema_consensus_timeout = 10000
2017/10/10 18:01:22 [debug] cassandra_ssl = false
2017/10/10 18:01:22 [debug] cassandra_ssl_verify = false
2017/10/10 18:01:22 [debug] cassandra_timeout = 5000
2017/10/10 18:01:22 [debug] cassandra_username = "kong"
2017/10/10 18:01:22 [debug] client_body_buffer_size = "8k"
2017/10/10 18:01:22 [debug] client_max_body_size = "0"
2017/10/10 18:01:22 [debug] client_ssl = false
2017/10/10 18:01:22 [debug] custom_plugins = {"nginx-log","helloworld-hdr","redirect","requestid","external","requirejwt","xsrf","418","authredir","cache-until-events"}
2017/10/10 18:01:22 [debug] database = "postgres"
2017/10/10 18:01:22 [debug] db_cache_ttl = 3600
2017/10/10 18:01:22 [debug] db_update_frequency = 5
2017/10/10 18:01:22 [debug] db_update_propagation = 0
2017/10/10 18:01:22 [debug] dns_error_ttl = 1
2017/10/10 18:01:22 [debug] dns_hostsfile = "/etc/hosts"
2017/10/10 18:01:22 [debug] dns_no_sync = false
2017/10/10 18:01:22 [debug] dns_not_found_ttl = 30
2017/10/10 18:01:22 [debug] dns_order = {"LAST","SRV","A","CNAME"}
2017/10/10 18:01:22 [debug] dns_resolver = {}
2017/10/10 18:01:22 [debug] dns_stale_ttl = 4
2017/10/10 18:01:22 [debug] error_default_type = "text/plain"
2017/10/10 18:01:22 [debug] http2 = false
2017/10/10 18:01:22 [debug] latency_tokens = true
2017/10/10 18:01:22 [debug] log_level = "notice"
2017/10/10 18:01:22 [debug] lua_code_cache = "on"
2017/10/10 18:01:22 [debug] lua_package_cpath = ""
2017/10/10 18:01:22 [debug] lua_package_path = "?/init.lua;./kong/?.lua"
2017/10/10 18:01:22 [debug] lua_socket_pool_size = 30
2017/10/10 18:01:22 [debug] lua_ssl_trusted_certificate = "/usr/local/kong/ssl/server.crt"
2017/10/10 18:01:22 [debug] lua_ssl_verify_depth = 1
2017/10/10 18:01:22 [debug] mem_cache_size = "128m"
2017/10/10 18:01:22 [debug] nginx_daemon = "off"
2017/10/10 18:01:22 [debug] nginx_optimizations = true
2017/10/10 18:01:22 [debug] nginx_user = "nobody nobody"
2017/10/10 18:01:22 [debug] nginx_worker_processes = "auto"
2017/10/10 18:01:22 [debug] pg_database = "kong"
2017/10/10 18:01:22 [debug] pg_host = "postgres-94"
2017/10/10 18:01:22 [debug] pg_password = "******"
2017/10/10 18:01:22 [debug] pg_port = 5432
2017/10/10 18:01:22 [debug] pg_ssl = false
2017/10/10 18:01:22 [debug] pg_ssl_verify = false
2017/10/10 18:01:22 [debug] pg_user = "kong"
2017/10/10 18:01:22 [debug] prefix = "/usr/local/kong/"
2017/10/10 18:01:22 [debug] proxy_access_log = "logs/access.log"
2017/10/10 18:01:22 [debug] proxy_error_log = "logs/error.log"
2017/10/10 18:01:22 [debug] proxy_listen = "0.0.0.0:80"
2017/10/10 18:01:22 [debug] proxy_listen_ssl = "0.0.0.0:443"
2017/10/10 18:01:22 [debug] real_ip_header = "X-Real-IP"
2017/10/10 18:01:22 [debug] real_ip_recursive = "off"
2017/10/10 18:01:22 [debug] server_tokens = true
2017/10/10 18:01:22 [debug] ssl = true
2017/10/10 18:01:22 [debug] ssl_cert = "/usr/local/kong/ssl/server.crt"
2017/10/10 18:01:22 [debug] ssl_cert_key = "/usr/local/kong/ssl/server.key"
2017/10/10 18:01:22 [debug] ssl_cipher_suite = "modern"
2017/10/10 18:01:22 [debug] ssl_ciphers = "ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256"
2017/10/10 18:01:22 [debug] trusted_ips = {}
{"level":"fatal","method":"main","msg":"lookup dd-statsd on 127.0.0.11:53: no such host","time":"2017-10-10T18:01:23Z"}
2017/10/10 18:01:23 [notice] 41#41: using the "epoll" event method
2017/10/10 18:01:23 [notice] 41#41: openresty/1.11.2.4
2017/10/10 18:01:23 [notice] 41#41: built by gcc 6.3.0 (Alpine 6.3.0)
2017/10/10 18:01:23 [notice] 41#41: OS: Linux 4.9.36-moby
2017/10/10 18:01:23 [notice] 41#41: getrlimit(RLIMIT_NOFILE): 4096:4096
2017/10/10 18:01:23 [notice] 41#41: start worker processes
2017/10/10 18:01:23 [notice] 41#41: start worker process 49
2017/10/10 18:01:23 [notice] 41#41: start worker process 50
2017/10/10 18:01:23 [notice] 41#41: start worker process 51
2017/10/10 18:01:23 [notice] 41#41: start worker process 52
```
- Kong error logs (`<KONG_PREFIX>/logs/error.log`)
    + None
- Kong configuration (registered APIs/Plugins & configuration file)
```json
{
    "created_at": 1507659840000,
    "strip_uri": true,
    "id": "ef197831-08f7-4772-aa3d-95f961190678",
    "name": "example-api",
    "http_if_terminated": false,
    "https_only": false,
    "upstream_url": "http://localhost:8080",
    "uris": [
        "/example-api"
    ],
    "preserve_host": true,
    "upstream_connect_timeout": 60000,
    "upstream_read_timeout": 60000,
    "upstream_send_timeout": 60000,
    "retries": 5
},
{
    "created_at": 1507659840000,
    "strip_uri": true,
    "id": "04be0978-f98e-45f0-ae00-eda1050f283f",
    "name": "e",
    "methods": [
        "GET",
        "POST"
    ],
    "http_if_terminated": false,
    "https_only": false,
    "upstream_url": "http://localhost:8181",
    "uris": [
        "/e"
    ],
    "preserve_host": true,
    "upstream_connect_timeout": 60000,
    "upstream_read_timeout": 60000,
    "upstream_send_timeout": 60000,
    "retries": 5
}
```
- Operating System
    + Linux 36669050a8bb 4.9.36-moby #1 SMP Wed Jul 12 15:29:07 UTC 2017 x86_64 Linux
    + alpine-release: 3.6.2

Note: We've been able to get around this issue by setting both routes to have *explicit* allowed methods. It appears this issue _only_ occurs if one route has *implicit* methods and another route *explicit* methods.

Secondly - this issue occurs ONLY if the second route's URI is a subset of the first route. Let me write up the scenarios we've tested.

**Buggy behavior**:

Route 1: `/example-api`, Methods defined: None
Route 2: `/e`, Methods defined: `["GET"]`

**Not buggy behavior**:

Route 1: `/example-api`, Methods defined: `["GET"]`
Route 2: `/e`, Methods defined: `["GET"]`

**Not buggy behavior**:

Route 1: `/example-api`, Methods defined: None
Route 2: `/e`, Methods defined: None

**Not buggy behavior**:

Route 1: `/example-api`, Methods defined: None
Route 2: `/eb`, Methods defined: `["GET"]`

^^^ Note that this last example illustrates that the 2nd route's URI being a subset of the 1st route's URI has an effect on the buggy behavior.

A workaround (we have not fully tested yet), is to *always* define allowed methods for *all* API's (and not rely on implicit allowed methods).

Please let me know if I can provide any other info.

TIA

**Comment by thibaultcha** [comment]
Hi @dselans,

Thanks for the report. I was indeed able to reproduce on our latest release 0.11.1.

**Comment by thibaultcha** [comment]
Added to our internal tracking tools, we intend to fix this for our 0.12 release, or, potentially a 0.11.2 release, if 0.12 gets delayed too much. And as always: we also accept contributions... 😉 

Thanks!

**Comment by thibaultcha** [comment]
After spending some time and spawning a few reproducible test cases, I am afraid that I'll be backtracking on my previous statement that this was a buggy behavior, and actually affirm that what you are seeing is the intended behavior of the router. Sorry for not seeing it sooner. Explanation:

1. Let's re-affirm the postulate that when not specifying regexes in the `uris` field, matching is evaluated by *prefix*.
2. That is, if we have `uris=/e` and `uris=/example`, *both* would match  a given request if its URL is `/example`.
3. APIs are evaluated in a specific orders: when an API has more matching attributes than another, it is evaluated for matching before. An API with `uris+methods` will be evaluated before an API with only `uris`.
4. Hence, when you make the following request: `GET /example`, the router finds that the API with `uris=/e + methods=GET` is a perfectly valid match, and rightfully so.

It seems you have setup conflicting routes, at least for GET methods.

If you are in doubt regarding how the router works, or if you want to experiment with URI regexes to circumvent this mis-configuration, feel free to have a look at our [Proxying Guide](https://getkong.org/docs/0.11.x/proxy/), which covers the workings of the Kong's router.

I'll be closing this now, hoping to have shed some light on what is going on here. Thank you for the report, and sorry again for the confusion!
